students/YingGuo/lession07/html_render.py

```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python3
+
+"""
+A class-based system for rendering html.
+"""
+
+
+# This is the framework for the base class
+class Element():
+    tag_name = "html"
+    indent = "    "
+
+    def __init__(self, content=None, **kwargs):
+        if content is None:
+            self.content = []
+        if content is not None:
+            self.content = [content]
+        self.kwargs = kwargs 
+
+
+    def append(self, new_content):
+        self.content.append(new_content)
+
+    def write_html(self):
+        with open("output_html","w") as f:
+             f.write(self.content)
+             
+    #what does ind mean in the test_html? why it is needed?
+    def render(self, out_file, cur_ind=""):
+        if cur_ind:
+            out_file.write(cur_ind)
+        else:
+            pass
+
+        for content in self.content:
+            out_file.write("<{}>\n".format(self.tag_name))
+
+            if cur_ind:
+                out_file.write(cur_ind)
+            else:
+                pass
+            #
+            out_file.write(self.indent)
+            try:
+                content.render(out_file)
+            except AttributeError:
+                out_file.write(content)
+            out_file.write("\n")
+            
+            out_file.write("</{}>\n".format(self.tag_name))
+
+
+#a subclass of Element with tag body
+class Body(Element):
+    tag_name = "body"
+    indent = "    "*2
+
+#subclass for paragraph
+class P(Element):
+    tag_name = "p"
+    indent = "    "*3
+
+    def render(self, out_file, cur_ind=""):
+        if cur_ind:
+            out_file.write(cur_ind)
+        else:
+            pass
+
+        for content in self.content:
+            if self.kwargs is not None:
+                open_tag = ["<{}".format(self.tag_name)]
+                for x in self.kwargs:
+                    open_tag.append(x)
+                    open_tag.append("=")
+                    open_tag.append('"{}"'.format(self.kwargs[x]))
+                    open_tag.append(" ")
+                open_tag.append(">\n")
+                out_file.write("".join(open_tag))
```

Comment:
This attribute-rendering logic is needed everywhere, not just in the `p` element only, so this code should be moved up in the `Element` class.